classes/stream_wrapper.php

```diff
@@ -0,0 +1,542 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace local_azureblobstorage;
+
+use GuzzleHttp\Exception\ClientException;
+use GuzzleHttp\Psr7\CachingStream;
+use GuzzleHttp\Psr7\Utils;
+use local_azureblobstorage\api;
+use HashContext;
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * Azure Blob Storage stream wrapper to use "blob://<container>/<key>" files with PHP.
+ *
+ * Implementation references,
+ * https://github.com/aws/aws-sdk-php/blob/master/src/S3/StreamWrapper.php
+ * https://phpazure.codeplex.com/SourceControl/latest#trunk/library/Microsoft/WindowsAzure/Storage/Blob/Stream.php
+ *
+ * @package    local_azureblobstorage
+ * @author     Matthew Hilton <matthewhilton@catalyst-au.net>
+ * @copyright  Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class stream_wrapper {
+
+    /** @var resource|null Stream context (this is set by PHP) */
+    public $context;
+
+    /** @var StreamInterface Underlying stream resource */
+    private $body;
+
+    /** @var int Size of the body that is opened */
+    private $size;
+
+    /** @var string Mode in which the stream was opened */
+    private $mode;
+
+    /** @var string The opened protocol (e.g. "blob") */
+    private $protocol = 'blob';
+
+    /** @var HashContext Hash resource that is sent when flushing the file to Azure. */
+    private $hash;
+
+    /** @var bool records whether the file was readable when validating the stream_handle */
+    private $readable = true;
+
+    /** @var string The key extracted from the path from when the stream was opened */
+    private $key = null;
+
+    /**
+     * Register the blob://' stream wrapper
+     *
+     * @param api $client Client to use with the stream wrapper
+     * @param string $protocol Protocol to register as.
+     */
+    public static function register(api $client, $protocol = 'blob') {
```

Comment:
Should we ever allow anything OTHER than blob here? This feels like it should be hardcoded rather than a param

Reply:
Technically the protocol _is_ a param in the stream wrapper spec. If you changed it it would stop working with objectfs (objectfs would need to have it changed on its side as well) but technically this plugin is not objectfs specific, so I don't see a reason why it needs to be hardcoded.